### ISSUE TITLE
sharedMemExtern not working

### ISSUE THREAD

**Issue by mxmlnkn** [ComputationalRadiationPhysics/cupla#29]
original:

    extern __shared__ __align__( sizeof(T_PREC) ) unsigned char dynamicSharedMemory[];

converted version:

    sharedMemExtern( dynamicSharedMemory, unsigned char * );

gets me:

    cudaGaussian.tpp:808:9: error: invalid conversion from ‘unsigned char**’ to ‘unsigned char*’ [-fpermissive]

and just to try it:

    sharedMemExtern( dynamicSharedMemory, unsigned char * );

gets me:

    cudaGaussian.tpp:808:9: error: invalid conversion from ‘unsigned char*’ to ‘unsigned char’ [-fpermissive]

I think adding a pointer to this:
```C++
    #define sharedMemExtern(ppName, ...)   \
    __VA_ARGS__ /* fix -> */ *ppName =    \
        ::alpaka::block::shared::dyn::getMem<__VA_ARGS__>(acc)
```
should work, but I'm not sure if this is correct. But I can say, that it compiles and runs, but my tests fail with wrong results, but that could be some other porting error.

If this fix is correct, then also the `PortingGuide.md` needs to be corrected, because:

    extern __shared__ int * smInt;

becomes

    sharedMemExtern( smInt, int );

instead of

    sharedMemExtern( smInt, int * );

See commit https://github.com/mxmlnkn/cupla/commit/cd30ff81f0701325f793c956923429b18c7dc449

**Comment by mxmlnkn** [comment]
It most definitely is some problem with shared memory. After tracing the segfault I tried to test if shared memory works by adding a new line, which writes to the very first element. The segfault appeared there:
```C++
    template<class T_PREC>
    struct cudaKernelApplyKernelSharedWeights
    {
    template< class T_ACC >
    ALPAKA_FN_NO_INLINE_ACC
    void operator()
    (
        T_ACC const & acc,
        /* You can't pass by reference to a kernel !!! compiles, but gives weird errors ... */
        T_PREC * const rdpData,
        unsigned int const rImageWidth,
        T_PREC const * const rdpWeights,
        unsigned int const nKernelHalf
    ) const
    {
        /* manage dynamically allocated shared memory */
        /* @see http://stackoverflow.com/questions/27570552/ */
        sharedMemExtern( dynamicSharedMemory, unsigned char );
        T_PREC * const smBlock = reinterpret_cast<T_PREC*>( dynamicSharedMemory );

        __syncthreads();
        smBlock[0] = 0;
        // [...]
}
```

called with
```C++
        CUPLA_KERNEL( cudaKernelApplyKernelSharedWeights<T_PREC> )(
            nBlocks,nThreads,
            sizeof(T_PREC)*( kernelSize + bufferSize ),
            rStream
        )( rdpData, rnDataX, dpKernel, N );
```

`sizeof(T_PREC)*( kernelSize + bufferSize )` translates to `1028` so 1028 bytes of shared memory should be available.

If this doesn't help I could try to create a (non-)working minimal example. Maybe this is an alpaka bug, though. I'm still using the alpaka OpenMP backend, because CUDA doesn't work yet.

**Comment by psychocoderHPC** [comment]
extern shared memory was fixed with #25 but I miss to update the documentation. Docu fix is one the way.

**Comment by mxmlnkn** [comment]
Ah my bad, I should watch the updates more closely.

**Comment by psychocoderHPC** [comment]
No problem it was a little bit confuse because of the current broken cupla version. Travis-CI tests are on the way to solve all problems^^

This issue was closed with #25
